source/fab/errors.py

```diff
@@ -0,0 +1,344 @@
+        absent.  Used to format the error message.  Defaults to False.
+    """
+
+    def __init__(
+        self, api: Union[str, None], target: str, present: Optional[bool] = False
+    ) -> None:
+        self.target = target
+        self.present = present
+        self.api = api
+
+        message = "called "
+        if api:
+            message += f"with {api} API "
+        else:
+            message += "without API "
+        if present:
+            message += "and with "
+        else:
+            message += "but not with "
+        message += f"{target}"
+
+        super().__init__("psyclone", message)
+
+
+class FabToolNotAvailable(FabToolError):
+    """An unavailable tool has been requested.
+
+    Error where a tool which is not available in a particular suite of
+    tools has been requested.
+
+    :param tool: name of the current tool
+    :param suite: optional name of the current tool suite
+    """
+
+    def __init__(
+        self, tool: Union["Category", "Tool", str], suite: Optional[str] = None
+    ) -> None:
+        message = "not available"
+        if suite:
+            message += f" in suite {suite}"
+        super().__init__(tool, message)
+
+
+class FabToolInvalidSetting(FabToolError):
+    """An invalid tool setting has been requested.
+
+    Error where an invalid setting, e.g. MPI, has been requested for a
+    particular tool.
+
+    :param setting_type: name of the invalid setting
+    :param tool: the tool to which setting applies
+    :param additional: optional additional information
+    """
+
+    # FIXME: improve these here and in tool_repository
+
+    def __init__(
+        self,
+        setting_type: str,
+        tool: Union["Category", "Tool", str],
+        additional: Optional[str] = None,
+    ) -> None:
+        self.setting_type = setting_type
+
+        message = f"invalid {setting_type}"
+        if additional:
+            message += f" {additional}"
+
+        super().__init__(tool, message)
+
+
+class FabUnknownLibraryError(FabError):
+    """An unknown library has been requested.
+
+    Error where an library which is not known to the current Linker
+    instance is requested.
+
+    :param library: the name of the unknown library
+    """
+
+    def __init__(self, library: str) -> None:
+        self.library = library
+        super().__init__(f"unknown library {library}")
+
+
+class FabCommandError(FabError):
+    """An error was encountered running a subcommand.
+
+    Error where a subcommand run by the fab framework returned with a
+    non-zero exit code.  The exit code plus any data from the stdout
+    and stderr streams are retained to allow them to be passed back up
+    the calling stack.
+
+    :param command: the command being run
+    :param code: the command return code from the OS
+    :param output: output stream from the command
+    :param error: error stream from the command
+    """
+
+    def __init__(
+        self,
+        command: str,
```

Comment:
The subprocess module also allows commands to be specified as a list of strings. Check how it is done when this exception is raised - we might prefer to accept it straight rather than having to convert it.

In fact, looking below, I see you cover this eventuality so reflect it in the type hinting.